login.go

```diff
@@ -0,0 +1,227 @@
+	if err != nil {
+		return nil, "", err
+	}
+	vals := url.Values{}
+	vals.Add("client_id", clientID)
+
+	req, err := newReq(ctx, "POST", deviceCodeURL, strings.NewReader(vals.Encode()))
+	if err != nil {
+		return nil, "", err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, "", err
+	}
+	defer resp.Body.Close()
+
+	data, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, "", err
+	}
+
+	codeResp := &DeviceCodeResponse{}
+	err = json.Unmarshal(data, codeResp)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return codeResp, codeResp.VerificationURI, nil
+}
+
+func doTokenRequest(ctx context.Context, client *ConductoroneAPI, clientID string, deviceCodeResp *DeviceCodeResponse) (*tokenResponse, error) {
+	httpClient := client.sdkConfiguration.DefaultClient
+	baseURL := utils.ReplaceParameters(client.sdkConfiguration.GetServerDetails())
+
+	tokenURL, err := url.JoinPath(baseURL, "/auth/v1/token")
+	if err != nil {
+		return nil, err
+	}
+	vals := url.Values{}
+	vals.Add("client_id", clientID)
+	vals.Add("device_code", deviceCodeResp.DeviceCode)
+	vals.Add("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
+
+	startLoop := time.Now()
+	for {
+		select {
+		case <-time.After(time.Duration(deviceCodeResp.Interval) * time.Second):
+			if time.Now().After(startLoop.Add(time.Duration(deviceCodeResp.ExpiresIn) * time.Second)) {
+				return nil, errors.New("timeout")
+			}
+		case <-ctx.Done():
+			return nil, errors.New("context canceled")
+		}
+
+		req, err := newReq(ctx, "POST", tokenURL, strings.NewReader(vals.Encode()))
+		if err != nil {
+			return nil, err
+		}
+		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			return nil, err
+		}
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.StatusCode >= 300 {
+			errResp := &oauth2Error{}
+			err = json.Unmarshal(body, errResp)
+			if err != nil {
+				return nil, err
+			}
+
+			if errResp.ErrorType == "authorization_pending" {
+				continue
+			}
+
+			return nil, errors.New(errResp.ErrorDescription)
+		}
+
+		tokenResp := &tokenResponse{}
+		err = json.Unmarshal(body, tokenResp)
+		if err != nil {
+			return nil, err
+		}
+
+		return tokenResp, nil
+	}
+}
+
+func doClientCredentialRequest(ctx context.Context, client *ConductoroneAPI, tokenResp *tokenResponse, personalClientCredentialDisplayName string) (*clientResp, error) {
+	httpClient := client.sdkConfiguration.DefaultClient
+	baseURL := utils.ReplaceParameters(client.sdkConfiguration.GetServerDetails())
+
+	pccURL, err := url.JoinPath(baseURL, "/api/v1/iam/personal_clients")
+	if err != nil {
+		return nil, err
+	}
+
+	personalClientReq, err := newReq(ctx, "POST", pccURL, bytes.NewReader([]byte(fmt.Sprintf("{\"display_name\": \"%s\"}", personalClientCredentialDisplayName))))
```

Comment:
nit: this won't be property escaped.

should just make an local struct with json tags?